### DISCUSSION TITLE
content import into player UI

### DISCUSSION THREAD

**Discussion by TomWoodward** [amp-up-io/qti3-item-player-controller#9] [Ideas]
I have [stubbed out a PR](https://github.com/amp-up-io/qti3-item-player-controller/pull/6) that loads a qti package into javascript memory and allows you to navigate through the items in the player's test UI

I did this because i wanted to explore the player's capabilities with different example content, especially from the qti-examples repository. I found the subbed functionality very useful for this purpose already.

the functionality in place does the bare minimum to import items in a usable way, and does not approach capturing the requirements of the assessment spec with sections and parts and tests and stuff. I don't necessarily think it _needs_ to provide provide all that functionality to be useful, but it could be confusing if people expect it to.

there is a bigger question about the extent to which this project should try to encapsulate the full test logic. the reason i felt comfortable ignoring test/section grouping/progression in my data loader is that the current TestFactory utility doesn't support it anyway, it only has a flat list of items.

there are probably also parts of the individual item that my PR doesn't handle, i'm interested in this comment:
> Content Package import is an extremely advanced capability.  There are
mountains of code associated with CP Import and Item + Test Validation in
the Amp-up.io Content Bank.

is there code in the project already to handle import? i don't see any or anything called `the Amp-up.io Content Bank` anywhere, unless you mean the "QTI 3 Item Samples" in the sandbox

**Comment by paulgrudnitski** [comment]
There's a lot in your post to unpack!

First and foremost, THANK YOU for your contribution(s) and for digging into the two codebases (Player component and Player Controller app).

By chance, will you be at Learning Impact in Salt Lake City in a couple of weeks?  If so, would love to chat w/you.

In the meantime...

There are two additional private repo's that might be of interest to you.  Both of these repo's will be public and MIT licensed before the end of the summer (or soon thereafter).

1) A **QTI 3 Test component** that will ingest a `<qti-assessment-test>` and then provide an information model (test properties, test parts, sections, items, item session control, outcome processing, branching, timers, rubric blocks, etc.) about the test.

2) A **QTI 3 Test Player app** (we call this the "Test Explorer") that encapsulates both the QTI 3 Test component and the QTI 3 Item Player component to provide full rendering, scoring, and evaluation of the QTI 3 test, section, item models.

These two chunks of code are already in production - more or less - in the Amp-up Content Bank as well as in the Amp-up LTI connector to the Content Bank for Test rendering.

Some snapshots of the Test Explorer from the "S3-S4 Section - Selection Ordering" QTI 3 conformance test suite:

Test info:

![image](https://github.com/amp-up-io/qti3-item-player-controller/assets/898605/b2252b4e-2c0c-4c46-8c3e-126ce8659ae1)

Test Part info:

![image](https://github.com/amp-up-io/qti3-item-player-controller/assets/898605/27cf0506-f8d9-4b53-bc73-fa43cc156d61)

Section info:

![image](https://github.com/amp-up-io/qti3-item-player-controller/assets/898605/773e35cd-03e7-4155-9e31-b64959e4a7ce)

Item rendering:

![image](https://github.com/amp-up-io/qti3-item-player-controller/assets/898605/c4c908d4-9eb2-4c6a-9a15-efd2b7a78d42)

End of Section summary:

![image](https://github.com/amp-up-io/qti3-item-player-controller/assets/898605/ff9cc0f7-b96f-4e44-893a-7c852b1a8549)

End of Part summary, including scoring:

![image](https://github.com/amp-up-io/qti3-item-player-controller/assets/898605/af919bdf-83c3-4f45-9133-82f1ef4a7dc6)

Anyway...the code is very real and perhaps this will address some of your requirements.  I am in the process of converting everything to Vue 3, as well as documenting everything, so it will be a while before I release it.

**Comment by TomWoodward** [comment]
@paulgrudnitski thats very exciting! i look forward to playing with the new test components!

i won't be at Learning Impact this year, I have been attending the LTI workgroups for a few months and just started listening in on the QTI workgroup, by all accounts Learning Impact might be a good thing for me to plan on going to next time around.

regardless, i would love to chat if you want to over zoom or whatever. I'm interested in collaborating on an open source player but i have some requirements that might be faster to work out over voice.
